LitOptions class to hold backend options

GSterbrant opened the pull request:
### Description

Fixes #4871. 

The backend options for the lit shader are now defined in their own class, which makes it possible to know what options are available. 

Additionally, this allows us to validate options which might have moved to the backend object, and inform developers if they are still expecting the option to be there. This primarily pertains to options set through the onUpdateShader callback.

mvaligursky commented:
The docs for these should probably move to appropriate classes too
https://github.com/playcanvas/engine/blob/636571ee71f5257ec6a38f717e765401e2de374b/src/scene/materials/standard-material.js#L495-L564

mvaligursky commented:
And also, if the option has moved to litOptions, we should generate some deprecated properties that warn user the option has moved.

GSterbrant commented:
I added an options class which is used to hookup deprecation callbacks, and replaced the use of an anonymous options object with this new (albeit empty) one.

mvaligursky commented on src/deprecated/deprecated.js:
```diff
@@ -785,6 +787,24 @@ _defineAlias('glossVertexColor', 'glossMapVertexColor');
+    Object.defineProperty(Options.prototype, name, {
+        get: function () {
+            Debug.deprecated(`Getting pc.Options#${name} has been deprecated as the property has been moved to pc.Options.LitOptions.`);
+            return null;
```
could this return `this.LitOptions[name];` to be backwards compatible?

mvaligursky commented on src/scene/materials/lit-options.js:
```diff
@@ -0,0 +1,102 @@
+    BLEND_NONE, GAMMA_NONE
+} from '../constants.js';
+
+class LitOptions {
```
we should add the documentation to some of those properties that is still stored in the old place .. see my other comment on this PR.

GSterbrant replied:
Do you mean the properties stored in the ordinary options object?

mvaligursky commented on src/scene/materials/options.js:
```diff
@@ -0,0 +1,7 @@
+class Options {
```
we need to give it a less generic name than `Options`. Maybe `StandardOptions`?

GSterbrant replied:
Done

mvaligursky commented on src/scene/shader-lib/program-library.js:
```diff
@@ -95,7 +95,7 @@ class ProgramLibrary {
             const device = this._device;
             def = generator.createShaderDefinition(device, options);
-            def.name = `${name}-pass:${options.pass}`;
+            def.name = `${name}-pass:${options.litOptions?.pass}`;
```
I think .pass should stay on options as well to avoid changes in this file, as it needs to be compatible with other generators (for example basic.js) which has uses options.pass.

GSterbrant replied:
The shader backend needs access to the pass, the front end doesn't. The backend only has the backend options, so in that case we need to duplicate the pass. How do we solve that?

mvaligursky replied:
yeah duplication should be fine I think.

GSterbrant commented on src/deprecated/deprecated.js:
```diff
@@ -785,6 +787,24 @@ _defineAlias('glossVertexColor', 'glossMapVertexColor');
+            return null;
+        },
+        set: function (value) {
+            Debug.deprecated(`Setting pc.Options#${name} has been deprecated as the property has been moved to pc.Options.LitOptions.`);
```
Yeah, good catch, I've added that.

mvaligursky commented on src/scene/materials/lit-options.js:
```diff
@@ -0,0 +1,147 @@
+} from '../constants.js';
+
+/**
+ * - pass: value of {@link Layer#shaderPass} of the Layer being rendered. Must be set to the
+ * same in {@link StandardMaterialOptions#pass}.
```
Could we only have `pass` on StandardMaterialOptions, and the setter would also set it on LitOptions? Better than having to keep two in sync

mvaligursky commented on src/scene/materials/lit-options.js:
```diff
@@ -0,0 +1,147 @@
+ * - cubeMapProjection: the value of {@link StandardMaterial#cubeMapProjection}.
+ * - useMetalness: the value of {@link StandardMaterial#useMetalness}.
+ * - blendType: the value of {@link Material#blendType}.
+ * - twoSidedLighting: the value of {@link Material#twoSidedLighting}.
```
that would make them useable in typescript as well, as they would be correctly types

mvaligursky commented on src/scene/materials/standard-material-options.js:
```diff
@@ -0,0 +1,44 @@
+import { LitOptions } from "./lit-options.js";
+
+/**
+ * - pass: value of {@link Layer#shaderPass} of the Layer being rendered. Must be set to the
```
same here, document as properties.

mvaligursky commented on src/scene/materials/standard-material.js:
```diff
@@ -498,70 +498,10 @@ let _params = new Set();
- * - reflectionEncoding: one of null, "rgbm", "rgbe", "linear", "srgb"
- * - ambientSource: one of "ambientSH", "envAtlas", "constant"
- * - ambientEncoding: one of null, "rgbm", "rgbe", "linear", "srgb"
+ * assigned to the material, a reflection pass with simpler shaders and so on. These properties are
```
on top of this class is also this code

<img width="741" alt="Screenshot 2022-12-07 at 17 45 26" src="https://user-images.githubusercontent.com/59932779/206255311-8bc90338-a156-4e37-8002-7034120f4e4c.png">

we should replace both `*` by the options type.

mvaligursky replied:
This still needs to be done too.

GSterbrant replied:
Yep, fixed.

mvaligursky commented on src/scene/materials/lit-options.js:
```diff
@@ -0,0 +1,153 @@
+        this.hasTangents = false;
+        this.chunks = [];
+
+        this.pass = 0;
```
I think you need variable this._pass
getter would return it
setter would complain
and from outside, you'd do options._pass = ... to actually assign to it

GSterbrant replied:
Fixed

mvaligursky commented on src/scene/materials/lit-options.js:
```diff
@@ -0,0 +1,153 @@
+ * @property {boolean} alphaTest Enable alpha testing. See {@link Material#alphaTest}.
+ * @property {boolean} alphaToCoverage Enable alpha to coverage. See {@link Material#alphaToCoverage}.
+ * @property {boolean} opacityFadesSpecular Enable specular fade. See {@link Material#opacityFadesSpecular}.
+ * @property {float} ambientSH If ambient spherical harmonics are used. Ambient SH replace prefiltered cubemap
```
some missing full stops at the end of the sentences.

mvaligursky commented on src/scene/materials/lit-options.js:
```diff
@@ -0,0 +1,153 @@
+        this.useMetalness = false;
+        this.useDynamicRefraction = false;
+
+        this.fog = 'none';
```
is there some constant for this 'none'?

GSterbrant replied:
Yes! Good catch.

mvaligursky commented on src/scene/materials/standard-material-options.js:
```diff
@@ -0,0 +1,51 @@
+        this.lightMapEncoding = 'linear';
+        this.packedNormal = false;
+
+        this.litOptions = new LitOptions();
```
this needs JSdoc so that it's exposed to documentation.
Please build the API documentation locally to make sure it all builds correctly and is linked up from StandardMaterial

mvaligursky commented on src/scene/materials/standard-material-options.js:
```diff
@@ -0,0 +1,51 @@
+class StandardMaterialOptions {
+    constructor() {
+        this.chunks = [];
+        this.pass = 0;
```
this needs to be _pass as well, as we have setter / getter.

GSterbrant commented on src/scene/materials/lit-options.js:
```diff
@@ -0,0 +1,153 @@
+
+        this.reflectionSource = null;
+        this.reflectionEncoding = null;
+        this.ambientSource = 'constant';
```
Only string values afaik, that's how it's done in the standard material options builder.

mvaligursky commented on src/scene/materials/lit-options.js:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    get pass() {
+        Debug.warn(`pc.LitOptions#pass should be accessed by its parent pc.StandardMaterialOptions.`);
```
this should be no warning and return this._pass
otherwise code using litOptions.pass always gets 0

GSterbrant replied:
Yeah good point, this would break the lit-shader for sure.

GSterbrant replied:
Fixed.

mvaligursky reviewed (approved):
Looks good to me. Please do test it with at least engine examples to make sure all is happy.
